app/controllers/concerns/canvas_imscc_support.rb

```diff
@@ -0,0 +1,40 @@
+module Concerns
+  module CanvasImsccSupport
+    extend ActiveSupport::Concern
+
+    included do
+      before_action :validate_token
+      skip_before_action :verify_authenticity_token
+      respond_to :json
+    end
+
+    protected
+
+    def validate_token
+      decoded_token = AuthToken.decode(encoded_token, nil, false)
+      lti_key = decoded_token[1]["kid"]
+      if application_instance = ApplicationInstance.find_by(lti_key: lti_key)
+        token = AuthToken.valid?(encoded_token, application_instance.lti_secret)
+        raise InvalidTokenError, "Unable to decode IMSCC jwt token" if token.blank?
+        raise InvalidTokenError, "Invalid IMSCC jwt token payload" if token.empty?
+        token[0]
+      else
+        raise Exceptions::InvalidImsccTokenError, "LTI key provided does not match any known application instances."
+      end
+    rescue JWT::DecodeError, Exceptions::InvalidImsccTokenError => ex
+      user_not_authorized("Unauthorized: Invalid token: #{ex}")
+    end
+
+    def encoded_token
```

Comment:
Is there a reason this code needs to be different than the jwt verification code in the jwt_token concern?

Reply:
Yes. Canvas sends a jwt signed with the lti secret. We only need to validate that the signature is good. We don't care about the payload.